Document pagination design changes based on review feedback (#1644)
opensearch-project/sql: docs/dev/Pagination-v2.md

Currently, SQL plugin does not provide SQL cursor syntax. However, the SQL REST endpoint can return result a page at a time. This feature is used by JDBC and ODBC drivers.

# Scope

> At this time, V2 engine supports pagination only for simple `SELECT * FROM <table>` queries without any other clauses like `WHERE` or `ORDER BY`.

Reviewer: I feel like we're going to have update this line of the design each time we push a PR.  
Maybe describe what the scope of this document contains.  They will have to look at release notes and read issues to know what has actually been pushed to repo.

The discussion below uses *under max_result_window* to refer to scenarios that can be implemented with simple retrieval API and *over max_result_window* for scenarios that require scroll API to implement.

## SQL Node Load Balancing

V2 SQL engine supports *sql node load balancing* &mdash; a cursor request can be routed to any SQL node in a cluster. This is achieved by encoding all data necessary to retrieve the next page in the `cursor_id` property in the response.

## Feature Design
To support pagination, v2 SQL engine needs to:
1. in REST front-end:
    1. Route supported paginated query to v2 engine for
        1. Initial requests,
        2. Next page requests.
    2. Fallback to v1 engine for queries not supported by v2 engine.
    3. Create correct JSON response from execution of paginated physical plan by v2 engine.
2. during query planning:
    1. Differentiate between paginated and normal query plans.
    2. Push down pagination to table scan.
    3. Create paginated physical plan from cursor id.
3. during query execution:
    1. Execute paginated physical plan.
4. in OpenSearch data source: 
    1. Support pagination push down.

>     2. Support other push down optimizations with pagination.

Reviewer: What about de-serialization and creating the cursor for the _next_ requests?

### Query Plan Changes

> All three kinds of requests &mdash; non-paged, initial page, or subsequent page  &mdash; are processed in the same way. Simplified workflow of query plan processing is shown below for reference.

Reviewer: ```suggestion
All three kinds of requests &mdash; non-paged, initial page, or subsequent page  &mdash; are processed at a high-level in the same way. Simplified workflow of query plan processing is shown below for reference.
```

Reviewer: We want to say that all the requests go through the same high-level steps, although the details/execution of each step may differ.

> All three kinds of requests &mdash; non-paged, initial page, or subsequent page  &mdash; are processed in the same way. Simplified workflow of query plan processing is shown below for reference.

Reviewer 2: ```suggestion
All three kinds of requests &mdash; non-paged, initial query, or subsequent query &mdash; are processed in the same way. Simplified workflow of query plan processing is shown below for reference.
```

    Optimized --> Physical : Implementor
  }
```

#### Unresolved Query Plan

Unresolved Query Plan for non-paged requests remains unchanged. 

> To support initial page requests, the `QueryPlan` class has a new optional field `pageSize`.

Reviewer 2: ```suggestion
To support initial query requests, the `QueryPlan` class has a new optional field `pageSize`.
```

    -UnresolvedPlan plan
    -QueryService queryService
  }
  class UnresolvedQueryPlan {
    <<UnresolvedPlan>>
  }
  QueryPlan --* UnresolvedQueryPlan
```

When `QueryPlanFactory.create` is passed initial query request, it:

> 1. Adds an instance of `Paginate` unresolved plan as the root of the unresolved query plan. 

Reviewer 2: ```suggestion
1. Adds an instance of `Paginate` unresolved plan as the root of the unresolved query plan.
```

  }

  state "Subsequent Query Request" As Sub {
    Cursor
  }
```

#### Logical Query Plan

There are no changes for non-paging requests.

> Changes to logical query plan to support initial pagination request:

Reviewer 2: ```suggestion
Changes to logical query plan to support Initial Query Request:
```

Processing of an Initial Query Request has few extra steps comparing versus processing a regular Query Request:
1. Query validation with `CanPaginateVisitor`. This is required to validate whether incoming query can be paged. This also activate legacy engine fallback mechanism.
2. `Serialization` is performed by `PlanSerializer` - it converts Physical Query Plan into a cursor, which could be used query a next page.

> 3Traversal of Physical Query Plan to get total hits, which is required to properly fill response to a user.

Reviewer: ```suggestion
3. Traversal of Physical Query Plan to get total hits, which is required to properly fill response to a user.
```

Reviewer 2: I think you can omit total hits

#### Serialization

All query plan nodes which are supported by pagination should implement [`SerializablePlan`](https://github.com/opensearch-project/sql/blob/f40bb6d68241e76728737d88026e4c8b1e6b3b8b/core/src/main/java/org/opensearch/sql/planner/SerializablePlan.java) interface. `getPlanForSerialization` method of this interface allows serialization mechanism to skip a tree node from serialization. OpenSearch search request objects are not serialized, but search context provided by the OpenSearch cluster is extracted from them.

```mermaid
sequenceDiagram
    participant PlanSerializer
    participant ProjectOperator
    participant ResourceMonitorPlan
    participant OpenSearchIndexScan

>     participant OpenSearchScrollRequest

Reviewer 2: a duplicate?
```suggestion
```

  ProjectOperator ->>+ ResourceMonitorPlan : getPlanForSerialization
    ResourceMonitorPlan -->>- ProjectOperator : delegate
  Note over ResourceMonitorPlan : ResourceMonitorPlan<br />is not serialized
  ProjectOperator ->>+ OpenSearchIndexScan : writeExternal
    alt First page
      OpenSearchIndexScan ->>+ OpenSearchScrollRequest : writeTo
        OpenSearchScrollRequest -->>- OpenSearchIndexScan : serialized request
    else Subsequent page
      OpenSearchIndexScan ->>+ OpenSearchScrollRequest : writeTo
        OpenSearchScrollRequest -->>- OpenSearchIndexScan : serialized request
    end
    Note over OpenSearchIndexScan : dump private fields
    OpenSearchIndexScan -->>- ProjectOperator : serialized
  ProjectOperator -->>- PlanSerializer : serialized
Note over PlanSerializer : Zip to reduce size
```

#### Deserialization

> Deserialization restores previously serialized Physical Query Plan. The recovered plan is ready to execute and should return the next page of the search response. To complete the query plan restoration, SQL engine should build a new request to the OpenSearch node. This request doesn't contain a search query, but it contains a search context reference &mdash; `scrollID`. To create a new `OpenSearchScrollRequest` object it is required to access to the instance of `OpenSearchStorageEngine`. `OpenSearchStorageEngine` can't be serialized, and it exists as a singleton in the SQL plugin engine. `PlanSerializer` creates a customized deserialization binary object stream &mdash; `CursorDeserializationStream`. This stream provides an interface to access the `OpenSearchStorageEngine` object.

Reviewer: ```suggestion
Deserialization restores previously serialized Physical Query Plan. The recovered plan is ready to execute and returns the next page of the search response. To complete the query plan restoration, SQL engine will build a new request to the OpenSearch node. This request doesn't contain a search query, but it contains a search context reference &mdash; `scrollID`. To create a new `OpenSearchScrollRequest` object it requires access to the instance of `OpenSearchStorageEngine`. Note: `OpenSearchStorageEngine` can't be serialized, and it exists as a singleton in the SQL plugin engine. `PlanSerializer` creates a customized deserialization binary object stream &mdash; `CursorDeserializationStream`. This stream provides an interface to access the `OpenSearchStorageEngine` object.
```
